### PR TITLE
Tar: Improve unseekable stream handling

### LINKED ISSUE #76690: Improve handling when setting an unseekable stream as a tar entry's DataStream
Original report here: https://github.com/dotnet/sdk-container-builds/issues/192#issuecomment-1269007374

If the user creates a new regular file tar entry, and sets the `DataStream` to an unseekable stream, we should either:

- Throw immediately if the stream is unseekable in the [TarEntry.cs#L251 setter](https://github.com/dotnet/runtime/blob/221717bc1a48f164c5eb2f8ee050f04510a21d41/src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs#L251), or...
- Find a way to retrieve the unseekable stream's length and position without calling Length, right before writing the data to the archive: [TarHeader.Write.cs#L510-L511](https://github.com/dotnet/runtime/blob/221717bc1a48f164c5eb2f8ee050f04510a21d41/src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs#L510-L511)

cc @Jozkee

### PR DESCRIPTION
**Pull request by carlossanlop**
Fixes https://github.com/dotnet/runtime/issues/76690

Addresses the issue reported in https://github.com/dotnet/sdk-container-builds/issues/192#issuecomment-1269049141 by @rainersigwald

Before this change, it wasn't possible to write into an archive a file entry that contained an unseekable data stream. The undelying cause was that we were calling `Length` to try to get the total number of bytes of the data stream, but it is unsupported when the stream's `CanSeek` is `false`.

The way I fixed this was by separating the code that writes an entry into two possible paths:

- One path for entries with seekable data streams, which remains mostly unmodified.

- A new path for entries with unseekable streams, which changes the order in which we collect the entry information into the buffer:
  - First, write into the buffer all the metadata except for the size field.
  - Then, write the data stream into the archive (this is possible because we know the exact location where the data should start), counting the number of bytes that are written.
  - Afterwards, write that collected length value in the location for the size field in the buffer.
  - Finally, calculate the checksum, then write the buffer into the archive.

### CODE REVIEW

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs
Comment (adamsitnik) [inline src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:757]: Why allocate an array of size X (`TarHelpers.RecordSize`) just to slice it to size Y (`paddingAfterData`)?

```suggestion
                byte[] zeros = new byte[paddingAfterData];
                
                return archiveStream.WriteAsync(zeros, cancellationToken);
```

I assume that `paddingAfterData` is small and renting an array would not be beneficial.

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs
Comment (adamsitnik) [inline src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:746]: The method seems to be just reserving the space for padding by writing zero bytes. I would rename it to make it more obvious

```suggestion
        private static ValueTask WriteEmptyPaddingAsync(Stream archiveStream, long actualLength, CancellationToken cancellationToken)
```

File: src/libraries/System.Formats.Tar/tests/TarFile/TarFile.ExtractToDirectory.Stream.Tests.cs
Comment (adamsitnik) [inline src/libraries/System.Formats.Tar/tests/TarFile/TarFile.ExtractToDirectory.Stream.Tests.cs:234]: these names don't tell me much. They could be renamed to express their job

```suggestion
            byte[] fileContent = new byte[bufferLength];
            byte[] dataStreamContent = new byte[bufferLength];
```

**Review by adamsitnik — COMMENTED** [review]
Overall it LGTM, but it would be nice if we could reduce the code duplication before merging.

Thank you @carlossanlop !

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs
Comment (carlossanlop) [reply src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:49]: Good catch. I moved them all to a dedicated method as suggested. I actually prefer the switch when checking the different possible values of the same variable.

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs
Comment (carlossanlop) [reply src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:107]: I renamed the method to WriteFromUnseekableDataStream as suggested and also added the extra debug assert check to ensure the data stream is unseekable when calling this method.

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs
Comment (carlossanlop) [reply src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:104]: Good point. Although I prefer to put it in its own separate debug assert, so that when/if it fails, it's quick and easy to determine exactly which condition failed.

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs
Comment (carlossanlop) [reply src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:107]: Good catch because the async version of this method does include the `Data` word in the name.

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs
Comment (carlossanlop) [reply src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:231]: Ah sorry, I thought I had removed it. I originally had a separate write method for each format, but then I ended up merging them all into a single one. Deleting.

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs
Comment (carlossanlop) [reply src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:757]: We're not supposed to use variable allocation lengths when creating stackallocs.

Although the CalculatePadding method will return a number under 512, if that method ever changes its logic unexpectedly and stops having that upper limit, we could get in trouble.

There is a debug assert that confirms that paddingAfterData is smaller than RecordSize, but still.

BTW this was actually very nicely explained by our very own @vcsjones in his blog post https://vcsjones.dev/stackalloc/

Comment (adamsitnik) [reply src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:685]: > when creating stackallocs

That is correct, but this method is async and you are not using `stackalloc` but `new`.

**Review by adamsitnik — APPROVED** [review]
LGTM, please apply my suggestions before merging.

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs
Comment (carlossanlop) [reply src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:44]: Ah, right. My bad. That makes more sense. Fixing.

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs
Comment (carlossanlop) [reply src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:685]: Ah good point. I was looking at the sync one.

Comment (carlossanlop) [reply src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:685]: I assume the same rule of avoiding using a variable length does not apply when creating a byte array?

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs
Comment (adamsitnik) [inline src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:680]: nit: this needs a rename to fix the compiler error

```suggestion
                return archiveStream.WriteAsync(zeros, cancellationToken);
```

**Comment by carlossanlop** [comment]
Failures are unrelated.

**Pull request merged**
